Change ledstrip_profile to ledstrip_mode, and status to default

ctzsnooze opened the pull request:
When experimenting with LEDstrips and the new `RACE` mode from [PR 13096](https://github.com/betaflight/betaflight/pull/13096), I found it a bit confusing reverting the change of mode.

The requirement for normal LED mode - where the user makes settings on the LED Tab in Configurator, and they work - is that the CLI is `set ledstrip_profile = status`.  

I found this a bit counter-intuitive.  First, I wondered if this led strip 'profile' had something to do with other profiles.  Second, 'status' didn't imply the 'default' or 'normal' behaviour.  So then I felt a bit lost.  

This PR changes only the names:
- `ledstrip_profile` -> `ledstrip_mode`
- (ledstrip) `status` -> `default`.

To return to the default behaviour, the user would now go `set ledstrip_mode = default`. 

Although it's late, to be honest, given that we have some serious LED strip improvements in 4.5, I would very much like to make these changes now.  I'm concerned about user errors where they make the mode change and then find that nothing in the LED Strip configurator tab works anymore.  There is no warning, you see, that the RACE mode is configured.

No explicit Configurator or Blackbox changes are needed.

On the other hand, if people think it should wait for 4.6, that's fine too.

KarateBrot commented:
Renaming from "profile" to "mode" makes sense to me. But I really don't like the renaming from `STATUS` to `DEFAULT` because it is less descriptive and if we were to exclude the `USE_LED_STRIP_STATUS_MODE` macro, the `DEFAULT` mode would disappear and would no longer be the default, which is a bit illogical, imo.
LED profiles/modes have been introduced 5 years ago in PR https://github.com/betaflight/betaflight/pull/7303 and there were no issues. So I'm not gonna approve but maybe others think differently about it.

blckmn commented:
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> FAIL
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> FAIL
 - `Tested` label found -> PASS
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

ctzsnooze commented:
Much as I respect your thoughts, @haslinghuis and @KarateBrot, the issues are going to be support problems.

The concept that the normal way of operating is 'status' mode, when the normal mode of operating does *NOT* return `status` of anything, is, well, to speak frankly, not just stupid, it's misleading and confusing.

Maybe 5 years ago the 'status' mode only did that.  But at some point, the 'normal' LED options - those in the LED page - because very diverse, so that the LED setup page did far more than only show the 'status' of something.

Now, the options in the LED page can be constant colour, rainbow, warnings, various things.  Many options have nothing to do with the 'status' of anything at all.

So it is a misleading name now.

And also it is counter-intuitive, because of the three options, which is the 'normal' one?  It's not clear at all.

If we do not change `STATUS` to `DEFAULT`, then I am certain we will get user complaints that they tried the RACE Led mode, and then suddenly their LED configuration is broken - changes they make in the LED screen don't work - and they don't know how to fix it.

There is zero downside to making this change now.  Just because the name `STATUS` is old is not a good enough reason to keep  a misleading and counter-intuitive name.

`DEFAULT`, on the other hand, is truly the default.  And it is intuitively obvious that it is the default.

Finally, the proposed change is a name change only and is risk free.  It's long overdue.  We are making a number of changes to LED things at this point, so now is a good time to fix this issue.

ctzsnooze commented on src/main/cms/cms_menu_ledstrip.c:
```diff
@@ -61,7 +61,7 @@ const char * const ledProfileNames[LED_PROFILE_COUNT] = {
     "BEACON",
 #ifdef USE_LED_STRIP_STATUS_MODE
-    "STATUS"
+    "DEFAULT"
```
@haslinghuis the define `USE_LED_STRIP_STATUS_MODE` is always true in current code.  

A full revision of the PR to change the internal labels to being internally consistent, from say `LED_PROFILE_STATUS` to `LED_MODE_DEFAULT`, and  `USE_LED_STRIP_STATUS_MODE` to `USE_LED_STRIP_DEFAULT_MODE`, etc, would be a bigger undertaking and carry greater risk than just changing the names.

I'm happy to do that later, but just changing the user-name in the CLI is low risk and useful and should go ahead now, before 4.5 is released.

haslinghuis replied:
It would be simple for me :)

haslinghuis replied:
Another issue is the fact we now use USE_LED_STRIP_STATUS_MODE define for a DEFAULT declaration.

This needs to be fixed.

ctzsnooze replied:
What do you mean by 'needs to be fixed'?  
The PR code works properly, as is, when tested.  
Do you mean that you want me to change the name of `USE_LED_STRIP_STATUS_MODE` to `USE_LED_STRIP_DEFAULT_MODE` in this PR?

ctzsnooze replied:
I was trying to avoid more extensive changes.
The intent of this PR is purely user interface clarity, not a change to underlying code.

haslinghuis replied:
It's just about consistency. But having a DEFAULT item under a define seems bad coding.

ctzsnooze replied:
We probably would have to also change:
`PG_LED_STRIP_STATUS_MODE_CONFIG`,
`ledStripStatusModeConfig_t`,
`*ledStripStatusModeConfig`,
`ledStripStatusModeConfigMutable`

ctzsnooze commented:
@haslinghuis @KarateBrot I added a second, large commit that changes the names of all the LED related files in the code to use 'default' not 'setup'.  Tested; works fine.  May break unit tests, waiting on them.
If it breaks unit tests, than perhaps I should put those changes into a separate PR.
Otherwise yes, the code is now internally consistent throughout.

haslinghuis commented on src/main/io/ledstrip.h:
```diff
@@ -143,8 +143,8 @@ typedef enum {
     LED_PROFILE_BEACON,
-#ifdef USE_LED_STRIP_STATUS_MODE
-    LED_PROFILE_STATUS,
+#ifdef USE_LED_STRIP_DEFAULT_MODE
+    LED_PROFILE_DEFAULT,
```
This is still not right

ctzsnooze replied:
added a commit to change `LED_PROFILE_DEFAULT` and similar to `LED_MODE_DEFAULT`

haslinghuis replied:
Instead maybe we use should turn around and use `USE_LED_STRIP_RACE_MODE` so we use `LED_MODE_DEFAULT` by default and `LED_MODE_RACE` if `USE_LED_STRIP_RACE_MODE` was defined.

Please note the change from `PROFILE` to `MODE`.

ctzsnooze replied:
Does this solve the problem?
```
typedef enum {
    LED_USER_MODE_RACE = 0,
    LED_USER_MODE_BEACON,
#ifdef USE_LED_STRIP_DEFAULT_MODE
    LED_USER_MODE_DEFAULT,
#endif
    LED_USER_MODE_COUNT
} ledUserMode_e;
```
Note I can't use `LED_MODE_COUNT` here; it won't compile since there are internal LED modes with names like `LED_MODE_HORIZON`  that are counted with `LED_MODE_COUNT`.

KarateBrot commented:
I have a problem with the name `DEFAULT` because it does not describe the content of the ledstrip mode. Maybe `CUSTOM` (or something more fitting) is a better choice, cause then it hints at a mode in which you can select custom colors and patterns.

haslinghuis commented:
We should probably tie `USE_LED_STRIP_RACE_MODE` to `USE_PRO_RACE`

KarateBrot commented on src/main/io/ledstrip.c:
```diff
@@ -1276,7 +1276,7 @@ static void applySimpleProfile(timeUs_t currentTimeUs)
                 break;
 
-            case LED_PROFILE_BEACON: {
+            case LED_MODE_BEAON: {
```
There are a few accidental instances of `BEAON` (without "C")
```suggestion
            case LED_MODE_BEACON: {
```

ctzsnooze replied:
fixed, thanks.
However LED_MODE_COUNT was affected by usage of LED_MODE_ANGLE and LED_MODE_HORIZON, causing compile errors.
So changed these to LED_USER_MODE_xx to differentiate with functional modes.

haslinghuis commented on src/test/unit/target.h:
```diff
@@ -55,7 +55,7 @@
 #define USE_TELEMETRY_SMARTPORT
 #define USE_LED_STRIP
-#define USE_LED_STRIP_STATUS_MODE
+#define USE_LED_STRIP_DEFAULT_MODE
```
Removing `USE_LED_STRIP_DEFAULT_MODE` would make sense to the code changes.

haslinghuis commented:
Maybe we remove `USE_LED_STRIP_DEFAULT_MODE` all together as they would make more sense to the code changes as they would be covered by USE_LED_STRIP and the STATUS mode being accepted after all these years.

ctzsnooze commented:
The refactoring is really complex.

@KarateBrot I've changed the mode name `DEFAULT` to `LED_TAB` since functionally it activates the LED tab in Configurator.  The user can choose `BEACON`, `LED_TAB` or `RACE`.
Although I prefer `DEFAULT` I can live with this one.  
This implies the reality of either one or the other.

@haslinghuis  The define that enables the LED tab in Configurator is now `USE_LED_STRIP_LED_TAB`.  This mirrors the intent of those code blocks, which is to give the user access to the functionality in Configurator's LED tab.

I've squashed all the above commits into one to simplify reviewing.  That may cause us to lose some comments, and require re-review, but that's probably what we need to do.

KarateBrot commented:
I think `LED_TAB` is also not ideal beause the name doesn't make sense without the configurator UI in mind. The `STATUS` led profile/mode is not a tab after all. It has been working for 5 years and people got used to it. I just don't see why you need to do this right now at the possibly worst time, now that we are so close to RC1 (and actually overdue). Sorry but I'll abstain from approving this time.

nerdCopter commented:
> [..]] the name `DEFAULT` does not describe the content/behavior of the ledstrip mode. [...]

i agree on this point.
maybe `STATIC`, but i would not want to imply that it cannot not Larson, Rainbow, etc.. :man_shrugging:

ctzsnooze commented:
I have reverted to to the original simple version of the PR, replacing only the CLI representations:
- `ledstrip_profile` to `ledstrip_mode`
- ledstrip mode `STATUS` to `DEFAULT`

The reasoning remains the same as described at the start.

I will not refactor the internal code at this point. I can make another PR for that later. Although the refactoring is complete, and works, it introduces risks that we should not be entertaining at this point.

Whereas changing names now makes a lot of sense.

The problem that we need to deal with is that we have a new (4.5 only) `RACE` ledstrip mode, and if it is active, the Configurator's LED tab suddenly stops working, with no indication that the reason is that the mode is 'RACE` mode.

If, for example, a user loads a Preset that sets enables the `RACE` mode, and wants to go back to 'normal' LED configuration method, ie with the LED tab, they will find it broken.  To fix, they will need to go to the CLI and get out of `RACE` mode and go back to the `NORMAL` or `DEFAULT` way of configuring the ledStrip.  

Their problem will they face is that their auto-complete that the CLI options they see will be `BEACON`, `RACE`, and `STATUS`.  

None of these implies 'normal' or 'default' operation.  

`STATUS` is an explicitly misleading with ancient origins.  It is a name that I never heard of or knew about until I tested the `RACE` mode and Steve's LEDfix PR.

But the moment I did that testing, I realised that `STATUS` was not the right term to use to indicate the 'normal' manner of control.

Now I don't mind if we re-name the `ledstrip_mode` option that gets us back to the normal or default state either `NORMAL` or `DEFAULT` or `LED_TAB`, but we must not persist with `STATUS`.

The simple reason is that we will need a Preset, now people will be changing this to `RACE`, where we 'undo' the `RACE` setting.  And for that we will need a Preset that uses the name we decide to run with for `NORMAL` operation.  Whatever that is.  And this is a NEW issue in 4.5 because we never had `RACE` before 4.5.

We need to pick an intuitive name and to ditch `STATUS`.  We have to do this now.

Note that in both Beacon and Race modes, the LED tab stops working.  Currently, the `STATUS` option enables the Configurator Led Tab.

So of the three options `NORMAL` or `DEFAULT` or `LED_TAB`, I think that `LED_TAB` is probably best, because both Normal or Default could change in future to Race for instance.

Could everyone please focus on the best name to use in place of `STATUS`.  That's the only thing we need to focus on.

haslinghuis commented:
There is more to this.

If using `STATUS` for a long time now included in the build we just can remove the `USE_LED_STRIP_STATUS_MODE` define.
But it is not. It is only added for > 512K builds:

https://github.com/betaflight/betaflight/blob/03495f9273427c5203d7e6c9761692fcd275877e/src/main/target/common_pre.h#L281-L283

Default for 512K builds we have:
https://github.com/betaflight/betaflight/blob/03495f9273427c5203d7e6c9761692fcd275877e/src/main/target/common_post.h#L467-L469

If we would remove `USE_LED_STRIP_STATUS_MODE` we no longer need `USE_WS2811_SINGLE_COLOUR` code.

As `STATUS` does not describe the function anymore as it got extended - maybe something like `EXTENDED` would do.

Think we should look at configurator for mode selection in led tab [first].

ctzsnooze commented:
We need to only change the UI names, not the underlying code.  Refactoring the underlying code is a very big task.

I've accepted feedback and changed `STATUS` to `NORMAL`, not my original suggestion of `DEFAULT`.

For the user to get back to normal control with the LED tab, they choose normal.  Once they choose RACE, the LED page is ignored.

We should, in the 4.6 time-frame, update Configurator to set RACE and BEACON modes, and make it obvious that the other settings won't apply when they are active.  And we can refactor all the LED code if someone has time for it.

KarateBrot commented:
To me it starts to look like this is more of a problem of the configurator UI, not of the BF firmware. Maybe it's a good idea to work on the LED tab UI to reflect all the LED features that are possible to configure. For example making it possible to select the `ledstrip_profile/mode` via UI.

ctzsnooze commented:
@haslinghuis I'd like to 're-awaken' this PR if possible.
It's a simple one, and I think should be included in 4.6.
Please review again and approve if you're happy.

SteveCEvans commented on src/main/cms/cms_menu_ledstrip.c:
```diff
@@ -61,7 +61,7 @@ const char * const ledProfileNames[LED_PROFILE_COUNT] = {
     "RACE",
     "BEACON",
 #ifdef USE_LED_STRIP_STATUS_MODE
```
Should this not be `USE_LED_STRIP_NORMAL_MODE`?

ctzsnooze replied:
Fixed, thanks.

SteveCEvans commented on src/main/cms/cms_menu_ledstrip.c:
```diff
@@ -61,7 +61,7 @@ const char * const ledProfileNames[LED_PROFILE_COUNT] = {
     "RACE",
```
There are duplicate strings in `lookupTableLEDProfile`. Can we’re not use the same strings?

SteveCEvans commented on src/main/io/ledstrip.h:
```diff
@@ -175,7 +175,7 @@ typedef struct ledStripConfig_s {
     ioTag_t ioTag;
     ledStripFormatRGB_e ledstrip_grb_rgb;
-    ledProfile_e ledstrip_profile;
+    ledProfile_e ledstrip_mode;
```
Make this `ledMode_e`?

ctzsnooze commented:
@haslinghuis ...
> the USE_LED_STRIP_STATUS_MODE define... is included in all builds
>Default for 512K builds we have:
https://github.com/betaflight/betaflight/blob/03495f9273427c5203d7e6c9761692fcd275877e/src/main/target/common_post.h#L467-L469

As I understand it, `USE_LED_STRIP` alone includes the 'simple' single colour modes `RACE` and `BEACON` only, and hence `USE_WS2811_SINGLE_COLOUR` applies.

If, however, we *add* the `NORMAL`, or complicated multi-colour methods shown in Configurator, then `USE_LED_STRIP_NORMAL_MODE` is active as well as `USE_LED_STRIP`.

In most usage cases, the `USE_WS2811_SINGLE_COLOUR ` check only within `USE_LED_STRIP` defines, and for these, I agree that they could be replaced by `#IFNDEF USE_LED_STRIP_NORMAL_MODE`.  However there are other places where we would need to determine what to do based on both the status of `USE_LED_STRIP` and `USE_LED_STRIP_NORMAL_MODE`.  

The use of `USE_WS2811_SINGLE_COLOUR` does appear to simplify the readability of the code.

Should I try to get rid of `USE_WS2811_SINGLE_COLOUR`?  At the moment that seems to me to be in the too-hard basket.

ctzsnooze commented:
@SteveCEvans @haslinghuis 
I went through this code carefully and made LOTS of changes.  It was very confusing, but I hope that now it makes a bit more sense with a consistent naming structure.

The main three options that the user sees in the CLI are now referred to as the Led strip 'Modes', not Profiles.  So there's no confusion with PID profiles.  `LED_PROFILE_RACE` becomes `LED_MODE_RACE`. These three primary modes are mutually exclusive.

The two simple Modes are `RACE`, and `BEACON`, their names don't change.  At this point they don't have any Configurator GUI options, but can be enabled in the CLI.  We need to update Configurator GUI so that the user can easily choose between these primary LED operation modes.

The third Mode is the somewhat complicated mode, which used to be called 'STATUS`. It can be set in Configurator's LED Tab to reflect throttle, orientation, flight mode, address individual LEDs, layer warnings over the top, etc etc. 

I struggle with the name for this Mode.  It was called `STATUS`, I think, because it could reflect certain flight controller status parameters, eg throttle and so on.  But `RACE` mode reflects the 'status' of the VTX. We previously discussed 'default' and kind of agreed on 'normal'.  However I remained unhappy about that, because for me, the `RACE` mode is 'normal'.  And depending on FC MPU flash space, you may not get a 'normal' option, which would seem strange.

In the end I thought about `FULL` or `MASTER` for that mode, on the basis that it gives full, or master, levels of control.  And I settled on `MASTER`.  I don't particularly like 'master' but it was the best I could think of.

And wherever the code is working on a 'Master' mode functionality, the  name now includes 'master', so we know what it's doing.

The word Profile was also used within the code for stepwise sub-task execution of the different sub-parts of the code, so I tried to avoid that terminology, leading to the use of 'step' and 'update_step'.  So that `LED_PROFILE_SLOW` becomes `LED_STRIP_UPDATE_STEP_SLOW`, and `ledProfileSequence` becomes `ledStripUpdateStep`.

There was also an internal table called 'Mode', with a `mode_count`, which included a list of parameters such as the Flight Controller's FlightMode, its Orientation, and the like.  When in 'Master' mode, it's one of the set of options that the user can choose from.  I used `FCSTATE` for the name of that list, since the options all relate to things the Flight Controller knows about.  Its not ideal but avoids confusion in the code with the main operating Modes.  So that `LED_MODE_ORIENTATION` becomes `LED_FCSTATE_ORIENTATION`, and `LED_MODE_HORIZON` becomes `LED_FCSTATE_HORIZON`.  

I think this code needed a comprehensive re-naming of the internal states, and I hope that it's now more legible as a result.

haslinghuis commented:
@ctzsnooze 

> As I understand it, USE_LED_STRIP alone includes the 'simple' single colour modes RACE and BEACON only, and hence USE_WS2811_SINGLE_COLOUR applies.

USE_WS2811_SINGLE_COLOUR does not seem to be used: https://github.com/betaflight/betaflight/pull/13219#issuecomment-1853057894

Still thinking about a better name for STATUS.

nerdCopter commented:
presuming this requires configurator changes?

ctzsnooze commented:
@haslinghuis - I changed the name of `STATUS` mode to `DETAILED` mode.
Hopefully that works.
Rebased and ready for review.

haslinghuis commented on src/main/io/ledstrip.h:
```diff
@@ -146,20 +146,20 @@ typedef enum {
-    LED_PROFILE_STATUS,
+    LED_STRIP_MODE_RACE = 0,
+    LED_STRIP_MODE_BEACON,
+#ifdef USE_LED_STRIP_DETAILED_MODE
+    LED_STRIP_MODE_MASTER,
 #endif
```
```suggestion
    LED_STRIP_MODE_DETAILED,
```

ctzsnooze replied:
Another good find, thank you

haslinghuis commented on src/main/io/ledstrip.h:
```diff
@@ -92,15 +92,15 @@ typedef enum {
+    LED_FCSTATE_BARO,
+    LED_FCSTATE_SPECIAL,
+    LED_FCSTATE_AUX_CHANNEL
+} ledFCStates_e;
```
```suggestion
} ledStates_e;
```

ctzsnooze replied:
These states are purely related to events in the FC, for use only when the 'detailed mode' wants to indicate an FC state.  They are not more general led states, so they are identified as being specifically FC related.  I was following the naming strategy of, for instance, `ledBaseFunctionId_e` and `ledOverlayId_e`.  Hence I think this name should stay as it is.
I am open to a different identifier than 'FC' in the name, however.  Its a kind of flight_mode modifier, but actually it includes other things like Baro readings and Aux channels. I really wasn't sure exactly how to name it, but they were all flight-controller related, so I chose 'FC', despite, and I agree, it looking a bit weird.

sugaarK commented:
Will need to change the presets that use it…

ctzsnooze commented on src/main/cli/cli.c:
```diff
@@ -1954,14 +1954,14 @@ static void cliLed(const char *cmdName, char *cmdline)
 
     if (isEmpty(cmdline)) {
-        printLed(DUMP_MASTER, ledStripStatusModeConfig()->ledConfigs, NULL, NULL);
+        printLed(DUMP_MASTER, ledStripDetailedModeConfig_t()->ledConfigs, NULL, NULL);
```
OMG I did that in lots of places! Stupid of me - thanks!
I've gone through very carefully and am sure it's correct now.
Strange that it compiled and that the tests all passed.

ctzsnooze commented on src/main/cli/cli.c:
```diff
@@ -2011,13 +2011,13 @@ static void cliColor(const char *cmdName, char *cmdline)
 }
 
-static void printModeColor(dumpFlags_t dumpMask, const ledStripStatusModeConfig_t *ledStripStatusModeConfig, const ledStripStatusModeConfig_t *defaultLedStripConfig, const char *headingStr)
+static void printModeColor(dumpFlags_t dumpMask, const ledStripDetailedModeConfig_t_t *ledStripDetailedModeConfig_t, const ledStripDetailedModeConfig_t_t *defaultLedStripConfig, const char *headingStr)
```
yes, lots of places. carefully fixed now.

ctzsnooze commented:
@sugaarK yes I think that to avoid errors, any preset that used a custom config for LEDstrip would need to be updated.

ctzsnooze commented:
@haslinghuis thanks for noticing those mistakes.  All of them should be fixed now, I went through it very carefully.  It's strange that the code appeared to work properly, and passed all the tests, anyway.

haslinghuis commented on src/main/io/ledstrip.h:
```diff
@@ -30,12 +30,12 @@
+#define LED_DIRECTION_COUNT               6
+#define LED_BASEFUNCTION_COUNT           10
+#define LED_OVERLAY_COUNT                 7
+#define LED_SPECIAL_FCSTATE_COLOR_COUNT 11
```
```suggestion
#define LED_FCSTATE_SPECIAL_COLOR_COUNT 11
```

nerdCopter replied:
makes sense; such would match this code:
```
src/main/io/ledstrip.c:1333:    } else if (fcStatusIndex == LED_FCSTATE_SPECIAL) {
src/main/io/ledstrip.h:101:    LED_FCSTATE_SPECIAL,
src/main/cli/cli.c:2039:            cliDefaultPrintLinef(dumpMask, equalsDefault, format, LED_FCSTATE_SPECIAL, j, colorIndexDefault);
src/main/cli/cli.c:2041:        cliDumpPrintLinef(dumpMask, equalsDefault, format, LED_FCSTATE_SPECIAL, j, colorIndex);
```
so modifying it, would need to change here as well:
```
src/main/io/ledstrip.c:1334:        if (modeColorIndex < 0 || modeColorIndex >= LED_SPECIAL_FCSTATE_COLOR_COUNT)
src/main/io/ledstrip.h:38:#define LED_SPECIAL_FCSTATE_COLOR_COUNT 11
src/main/io/ledstrip.h:162:    uint8_t color[LED_SPECIAL_FCSTATE_COLOR_COUNT];
src/main/msp/msp.c:1758:        for (int j = 0; j < LED_SPECIAL_FCSTATE_COLOR_COUNT; j++) {
src/main/cli/cli.c:2032:    for (uint32_t j = 0; j < LED_SPECIAL_FCSTATE_COLOR_COUNT; j++) {
```

ctzsnooze replied:
changed all instances, thanks

haslinghuis reviewed (approved):
This looks much better now.

ledvinap commented on src/main/cms/cms_menu_ledstrip.c:
```diff
@@ -57,11 +57,11 @@ static uint8_t cmsx_ledBeaconArmedOnly;
-#ifdef USE_LED_STRIP_STATUS_MODE
-    "STATUS"
+#ifdef USE_LED_STRIP_DETAILED_MODE
+    "DETAILED"
```
```suggestion
    [LED_STRIP_MODE_DETAILED] = "DETAILED",
```

ctzsnooze replied:
Done!  Thanks

ledvinap commented on src/main/io/ledstrip.c:
```diff
@@ -105,7 +105,7 @@ static uint8_t previousProfileColorIndex = COLOR_UNDEFINED;
 #define LED_EXEC_TIME_SHIFT             7
 
-#define PROFILE_COLOR_UPDATE_INTERVAL_US 1e6  // normally updates when color changes but this is a 1 second forced update
+#define COLOR_UPDATE_INTERVAL_US 1e6  // normally updates when color changes but this is a 1 second forced update
```
1e6 is `double` constant, with possible unintended consequences
```suggestion
#define COLOR_UPDATE_INTERVAL_US ((unsigned)1e6)  // normally updates when color changes but this is a 1 second forced update
```

ctzsnooze replied:
thanks

ledvinap commented on src/main/io/ledstrip.c:
```diff
@@ -1321,23 +1321,23 @@ bool parseColor(int index, const char *colorConfig)
     if (colorIndex < 0 || colorIndex >= LED_CONFIGURABLE_COLOR_COUNT)
         return false;
-    if (modeIndex < LED_MODE_COUNT) {  // modeIndex_e is unsigned, so one-sided test is enough
+    if (fcStatusIndex < LED_FCSTATE_COUNT) {  // ledFCStates_e is unsigned, so one-sided test is enough
```
`enum`s may be signed (it's implementation-defined). 
```suggestion
    if (fcStatusIndex >= 0 && fcStatusIndex < LED_FCSTATE_COUNT) {
```

ctzsnooze replied:
thanks, again

ctzsnooze replied:
Failed the build test so for now am reverting this one.

ledvinap commented on src/main/io/ledstrip.c:
```diff
@@ -1321,23 +1321,23 @@ bool parseColor(int index, const char *colorConfig)
         if (modeColorIndex < 0 || modeColorIndex >= 1)
             return false;
-        ledStripStatusModeConfigMutable()->ledstrip_aux_channel = colorIndex;
+        ledStripDetailedModeConfigMutable()->ledstrip_aux_channel = colorIndex;
```
`ledstrip_aux_channel` is RC channel, but colorIndex in stored into it?!

ctzsnooze replied:
yes that's rather strange... I have no clue, to be honest.

ledvinap replied:
I'll look into it

ctzsnooze replied:
Thanks!

ledvinap commented on src/main/io/ledstrip.c:
```diff
@@ -1458,23 +1458,23 @@ static ledProfileSequence_t applySimpleProfile(timeUs_t currentTimeUs)
-        colorUpdateTimeUs = currentTimeUs + PROFILE_COLOR_UPDATE_INTERVAL_US;
-        return LED_PROFILE_ADVANCE;
+    if ((colorIndex != previousColorIndex) || (currentTimeUs >= colorUpdateTimeUs)) {
+        setStripColor((useCustomColors) ? &ledStripDetailedModeConfig()->colors[colorIndex] : &hsv[colorIndex]);
```
```suggestion
        setStripColor(useCustomColors ? &ledStripDetailedModeConfig()->colors[colorIndex] : &hsv[colorIndex]);
```

ctzsnooze replied:
thanks, done.

ledvinap commented on src/main/io/ledstrip.c:
```diff
@@ -1495,62 +1495,62 @@ void ledStripUpdate(timeUs_t currentTimeUs)
-        if (applyProfile) {
-            ledProfileSequence_t ledProfileSequence = LED_PROFILE_SLOW;
+        if (isModeApplied) {
+            ledStripUpdateStep_t ledStripUpdateStep = LED_STRIP_UPDATE_STEP_SLOW;
```
```suggestion
            ledStripUpdateStep_t ledStripUpdateStep;
```

ctzsnooze replied:
OK, done

ledvinap commented on src/main/io/ledstrip.c:
```diff
@@ -1495,62 +1495,62 @@ void ledStripUpdate(timeUs_t currentTimeUs)
                 }
 
                 default:
                     break;
```
```suggestion
                    ledStripUpdateStep = LED_STRIP_UPDATE_STEP_SLOW;
                    break;
```

ctzsnooze replied:
Thanks.

ledvinap replied:
(the idea is to set LED_STRIP_UPDATE_STEP_SLOW explicitly for default)

ledvinap commented on src/main/io/ledstrip.h:
```diff
@@ -92,15 +92,15 @@ typedef enum {
+    LED_FCSTATE_HORIZON,
+    LED_FCSTATE_ANGLE,
+    LED_FCSTATE_MAG,
+    LED_FCSTATE_BARO,
```
`#define LED_FCSTATE_COUNT          (LED_FCSTATE_BARO + 1)`

ctzsnooze replied:
OK, though I don't understand much of the handling of the `LED_FCSTATE_AUX_CHANNEL` FC State...

ctzsnooze commented:
@ledvinap - have implemented your suggestions, many of which I didn't fully understand, but for which I am grateful, none the less.  Not tested functionally yet.

ctzsnooze commented:
@KarateBrot @sugaarK @haslinghuis could we please get this one approved and merged before we do 4.6?
The main advantage is having consistent naming for the update to the LED tab in configurator.

ctzsnooze commented:
I've rebased to master.
@blckmn @KarateBrot @sugaarK @haslinghuis - could we please get this one approved and merged before we do 4.6?
It gives consistent naming, needed for the update to the LED tab in configurator.

haslinghuis commented:
Still think `STATUS` is okay but going forward imho `DETAILED` should be `ADVANCED` and we need update in [documentation](https://betaflight.com/docs/development/LedStrip)

sugaarK commented:
> I've rebased to master. @blckmn @KarateBrot @sugaarK @haslinghuis - could we please get this one approved and merged before we do 4.6? It gives consistent naming, needed for the update to the LED tab in configurator.

will need to update the presets that use it too

blckmn commented:
@ctzsnooze you wanna rebase this one?

blckmn commented:
Closing this pull request as it has been open for over 12 months without being merged. Thank you for your contribution. If this is still relevant, please feel free to reopen or submit a new pull request.
